### DISCUSSION TITLE
Workspace Warnings

### DISCUSSION THREAD

**Discussion by jukzi** [eclipse-platform/.github#169] [Show and tell]
For an [OOmphed ](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md)Platform workspace 
![image](https://github.com/eclipse-platform/.github/assets/51790620/5b0f99e7-b1f6-45d6-97dd-9e4fcec2b147)
 

| Date | Errors|Warnings | Info | Total |
|--------------|---|-------:|------:|-----:|
| 2023-11-09 | 0 | 16373 | 2838 | 19211
| 2023-12-01 | 2 | 13283 | 1385 | 14668
| 2024-02-09 | 0 |   5906 |     21 | 5927
| 2024-03-05 | 0 |   4586 |     15 | 4601
| 2024-06-24 | 0 |   4974 |     15|  4989
| 2024-07-10 | 0 |   4953 |     15|  4968
| 2024-10-28 | 0 |   4175 |       1|  4176
| 2024-11-26 | 2 |   4175 |       1|  4178

would be nice to get it reduced.

**Comment by merks** [comment]
Once one ignores a dozen warnings one generally ignores them all including new ones.

**Reply by Bananeweizen** [reply]
What I generally do in company projects, and what might help others:
* Problems View > Group By > Type. Makes it easy to recognize missing encoding and other simple to fix non Java and PDE things. Also shows Build path issues separate of compile errors, making you notice these have to be fixed first.
* Problems View > Filter > Add a text filter for one specific warning text, then fix all occurrences, then set that warning to error in all project preferences of this project. Afterwards remove the filter. Repeat until there are no more warnings.
* Once zero is reached, configure egit to disable committing in case of any warning.
* Side note: This is only reliable, if _everything_ is stored as project settings, and workspace settings become basically ineffective.

We have several projects in the 6 digit LOC range with zero warnings using that approach (and with way stricter error/warning settings than default). I'm aware that introducing this later on existing code bases is harder, but the "fix one _type_ of warning and then make it an error" approach can be applied any time and definitely avoids adding more of that accidentally again.

**Comment by jukzi** [comment]
here is a list of frequent problem markers:
| count | stripped description |
| ------ | ------ |
| 6445 | ` is deprecated` |
| 3360 | `raw type should be parameterized` |
| 2779 | `Javadoc: ` |
| 1467 | `Empty block should be documented` |
| 508 | `The expression is boxed` |
| 462 | `and marked for removal` |
| 338 | `Discouraged access: The type '')` |
| 289 | `The switch statement should have a default case` |
| 227 | `The expression is unboxed` |
| 201 | `Potential resource leak: '' may not be closed` |
| 193 | `Type safety: needs unchecked conversion to conform to` |
| 189 | `The library '' has not been specified on the bundle classpath.` |
| 185 | `is hiding a field` |
| 178 | `The value of the parameter is not used` |
| 178 | `Non-externalized string literal; it should be followed by //$NON-NLS-<n>$` |
| 171 | `Potential null pointer access:` |
| 131 | `Discouraged access: The method '')` |
| 121 | `Project '' has no explicit encoding set` |
| 106 | `overrides a deprecated method` |
| 104 | `Discouraged access: The field '')` |
| 91 | `Type safety: Unchecked invocation` |
| 90 | ` is deprecated since version` |
| 87 | `The parameter should not be assigned` |
| 76 | `Resource '' should be managed by try-with-resource` |
| 71 | `is never used` |
| 53 | `The declared exception IOException is not actually thrown` |
| 47 | `The compiler compliance specified is but a JRE  is used` |
| 41 | `There are no JREs installed in the workspace that are strictly compatible with this environment.` |
| 39 | `Referenced class '' is not on the plug-in classpath` |
| 35 | `Potential resource leak: '' may not be closed at this location` |
| 33 | `should be accessed directly` |
| 32 | `A minimum of 1 '' elements must be specified.` |
| 31 | `Type safety: Unchecked cast ` |
| 30 | `Plug-in reference '' cannot be resolved` |
| 28 | `Referenced identifier '' cannot be found` |
| 26 | ` is emulated by a synthetic accessor method` |
| 25 | `At least one of the problems in category '' is not analysed due to a compiler option being ignored` |
| 24 | `Discouraged access: The constructor '')` |
| 22 | `'' is missing version constraint information` |
| 17 | `Unknown referenced nature:` |
| 14 | `Unlikely argument type for equals:` |
| 13 | `Element ''.` |
| 12 | `Access to referenced class '' is discouraged` |
| 11 | `The parameter is hiding another local variable defined in an enclosing scope` |
| 11 | `Execution environment is lower than one of the plug-in's dependencies` |
| 9 | `.*should be tagged with @Override since it actually overrides a superclass method` |
| 8 | `Neither an available bundle nor the associated JRE export package ''` |
| 8 | `Attribute '' must be defined.` |
| 7 | `Unnecessary @SuppressWarnings` |
| 6 | `The value for attribute '' is not externalized.  The value must begin with %` |
| 6 | `There is no '' build entry and the project has Java compliance preferences set` |
| 6 | `Referenced resource '' cannot be found` |
| 5 | `The switch over the enum should have a default case` |
| 5 | `should also implement hashCode since it overrides Object.equals` |
| 4 | `Source folders '' share the same output folder but are targeted to different libraries.` |
| 4 | `Dead code` |
| 4 | `Access to extension point '' is discouraged ` |

**Comment by jukzi** [comment]
markers by top level project (first 3 dots):
| count  | project |
| ---  | --- |
| 5549 | org.eclipse.jdt |
| 4403 | org.eclipse.ui |
| 1856 | org.eclipse.jface |
| 1721 | org.eclipse.equinox |
| 1223 | org.eclipse.osgi |
| 591 | org.eclipse.e4 |
| 533 | org.eclipse.ltk |
| 521 | org.eclipse.core |
| 400 | org.eclipse.team |
| 372 | org.eclipse.compare |
| 237 | org.eclipse.pde |
| 218 | org.eclipse.swt |
| 181 | org.eclipse.debug |
| 144 | org.eclipse.update |
| 100 | org.eclipse.text |
| 95 | org.eclipse.help |
| 94 | org.eclipse.test |
| 62 | org.eclipse.search |
| 52 | org.eclipse.ua |
| 46 | org.eclipse.ant |
| 28 | org.eclipse.jsch |
| 27 | org.eclipse.emf |
| 21 | org.eclipse.releng |
| 16 | org.eclipse.ecf |
| 11 | org.eclipse.unittest |
| 7 | org.eclipse.platform-feature |
| 4 | org.eclipse.platform |
| 2 | org.eclipse.urischeme |
| 2 | org.eclipse.jdt-feature |
| 0 | converterJclMi |
| 1 | org.eclipse.tips |
| 1 | org.eclipse.sdk |
| 1 | org.eclipse.rcp |
| 1 | org.eclipse.images |
